### PR TITLE
Add object Section.

### PR DESCRIPTION
**Pull request by siarheishut**
Add object Section in order to reuse sets of figures without copying it. Add unit-tests for it.

### CODE REVIEW

File: src/figures.cpp [hunk 1]
```diff
@@ -132,3 +133,29 @@ Rectangle &Rectangle::SetHeight(double height) {
...
+void svg::Section::Render(std::ostream &out) const {
+  for (auto &figure : *figures_) {
+    std::visit([&out](auto &&figure) {
+      figure.Render(out);
+    }, figure);
+  }
+}
+
+svg::Section::Section(std::vector<Object> objects) {
+  figures_ = std::make_shared<std::vector<Object>>(std::move(objects));
```
Comment (foxfrikses) [inline src/figures.cpp:146]: : figures{make_shared}

File: tests/figures_tests.cpp [hunk 2]
```diff
@@ -480,94 +540,68 @@ TEST(TestDocument, TestDocument) {
...
-                      "fill=\"none\" stroke=\"none\" stroke-width=\"1\" />"),
+                      DEFAULT_CIRCLE DEFAULT_POLYLINE DEFAULT_CIRCLE
+                      DEFAULT_TEXT DEFAULT_RECTANGLE DEFAULT_POLYLINE
+                      DEFAULT_RECTANGLE),
       },
   };
 
   for (auto &[name, figures, want] : test_cases) {
-    svg::Document doc;
+    auto doc = svg::Document{};
```
Comment (foxfrikses) [inline tests/figures_tests.cpp:597]: strange

File: tests/figures_tests.cpp [hunk 3]
```diff
@@ -480,94 +540,68 @@ TEST(TestDocument, TestDocument) {
...
       doc.Add(std::move(figure));
     }
-
-    std::ostringstream ss;
+    auto ss = std::ostringstream{};
     doc.Render(ss);
-    auto got = ss.str();
 
-    EXPECT_EQ(want, got);
+    auto got = ss.str();
```
Comment (foxfrikses) [inline tests/figures_tests.cpp:604]: this too

File: include/svg/document.h [hunk 4]
```diff
@@ -8,8 +8,6 @@
 #include "figures.h"
```
Comment (foxfrikses) [inline include/svg/document.h:8]: variant

File: include/svg/figures.h [hunk 5]
```diff
@@ -2,6 +2,7 @@
 #define SVG_FIGURES_H_
 
 #include <cstdint>
+#include <memory>
 #include <optional>
 #include <ostream>
 #include <vector>
```
Comment (foxfrikses) [inline include/svg/figures.h:8]: variant

File: include/svg/figures.h [hunk 6]
```diff
@@ -143,6 +151,26 @@ class Rectangle : public Figure<Rectangle> {
   double height_ = 0;
 };
 
+class Section {
+ public:
+  friend class SectionBuilder;
+  void Render(std::ostream &out) const;
+
+ private:
+  Section(std::vector<Object> objects);
```
Comment (foxfrikses) [inline include/svg/figures.h:160]: explicit

File: include/svg/figures.h [hunk 7]
```diff
@@ -143,6 +151,26 @@ class Rectangle : public Figure<Rectangle> {
...
+};
+
+class SectionBuilder {
+ public:
+  SectionBuilder &Add(const Object &figure);
+  SectionBuilder &Add(Object &&figure);
+  Section Build();
+
+ private:
+  std::vector<Object> figures_;
```
Comment (foxfrikses) [inline include/svg/figures.h:172]: objects_

File: include/svg/figures.h [hunk 8]
```diff
@@ -143,6 +151,26 @@ class Rectangle : public Figure<Rectangle> {
...
+ private:
+  Section(std::vector<Object> objects);
+
+  std::shared_ptr<std::vector<Object>> figures_;
+};
+
+class SectionBuilder {
+ public:
+  SectionBuilder &Add(const Object &figure);
+  SectionBuilder &Add(Object &&figure);
```
Comment (foxfrikses) [inline include/svg/figures.h:168]: object

File: include/svg/figures.h [hunk 9]
```diff
@@ -143,6 +151,26 @@ class Rectangle : public Figure<Rectangle> {
...
+
+ private:
+  Section(std::vector<Object> objects);
+
+  std::shared_ptr<std::vector<Object>> figures_;
+};
+
+class SectionBuilder {
+ public:
+  SectionBuilder &Add(const Object &figure);
```
Comment (foxfrikses) [inline include/svg/figures.h:167]: object

File: src/figures.cpp [hunk 10]
```diff
@@ -132,3 +133,29 @@ Rectangle &Rectangle::SetHeight(double height) {
...
+}
+
+svg::SectionBuilder &svg::SectionBuilder::Add(const svg::Object &figure) {
```
Comment (foxfrikses) [inline src/figures.cpp:149]: object

File: src/figures.cpp [hunk 11]
```diff
@@ -132,3 +133,29 @@ Rectangle &Rectangle::SetHeight(double height) {
...
+  figures_.push_back(figure);
+  return *this;
+}
+
+svg::SectionBuilder &svg::SectionBuilder::Add(svg::Object &&figure) {
```
Comment (foxfrikses) [inline src/figures.cpp:154]: object

File: src/figures.cpp (hunk 1 above)
Comment (siarheishut) [reply src/figures.cpp:146]: Done.

File: tests/figures_tests.cpp (hunk 2 above)
Comment (siarheishut) [reply tests/figures_tests.cpp:597]: Fixed.

File: tests/figures_tests.cpp (hunk 3 above)
Comment (siarheishut) [reply tests/figures_tests.cpp:604]: Fixed

File: include/svg/document.h (hunk 4 above)
Comment (siarheishut) [reply include/svg/document.h:7]: Done.

File: include/svg/figures.h (hunk 5 above)
Comment (siarheishut) [reply include/svg/figures.h:9]: Done.

File: include/svg/figures.h (hunk 6 above)
Comment (siarheishut) [reply include/svg/figures.h]: Done.

File: include/svg/figures.h (hunk 7 above)
Comment (siarheishut) [reply include/svg/figures.h]: Done.

File: include/svg/figures.h (hunk 9 above)
Comment (siarheishut) [reply include/svg/figures.h]: Done.

File: src/figures.cpp (hunk 10 above)
Comment (siarheishut) [reply src/figures.cpp]: Done.

File: src/figures.cpp (hunk 11 above)
Comment (siarheishut) [reply src/figures.cpp]: Done.

**Review by foxfrikses — APPROVED** [review]
LGTM

**Pull request merged**